src/content/services/monetizationLinkManager.ts

```diff
@@ -0,0 +1,498 @@
+                void this.run();
+              }
+            },
+            { once: true },
+          );
+        }
+      },
+      { once: true },
+    );
+  }
+
+  end() {
+    this.documentObserver.disconnect();
+    this.monetizationLinkAttrObserver.disconnect();
+    this.monetizationLinks.clear();
+    this.document.removeEventListener(
+      'visibilitychange',
+      this.onDocumentVisibilityChange,
+    );
+    this.window.removeEventListener('message', this.onWindowMessage);
+  }
+
+  /**
+   * Check if iframe or not
+   */
+  private async run() {
+    this.document.addEventListener(
+      'visibilitychange',
+      this.onDocumentVisibilityChange,
+    );
+
+    if (!this.isTopFrame && this.isFirstLevelFrame) {
+      this.window.addEventListener('message', this.onWindowMessage);
+    }
+
+    this.document
+      .querySelectorAll<HTMLElement>('[onmonetization]')
+      .forEach((node) => {
+        this.dispatchOnMonetizationAttrChangedEvent(node);
+      });
+
+    this.documentObserver.observe(this.document, {
+      subtree: true,
+      childList: true,
+      attributeFilter: ['onmonetization'],
+    });
+
+    const monetizationLinks = getMonetizationLinkTags(
+      this.document,
+      this.isTopFrame,
+    );
+
+    for (const link of monetizationLinks) {
+      this.observeLinkAttrs(link);
+    }
+
+    const validLinks = (
+      await Promise.all(monetizationLinks.map((elem) => this.checkLink(elem)))
+    ).filter(isNotNull);
+
+    for (const { link, details } of validLinks) {
+      this.monetizationLinks.set(link, details);
+    }
+
+    await this.sendStartMonetization(validLinks.map((e) => e.details));
+  }
+
+  private onWindowMessage = (event: MessageEvent<ContentToContentMessage>) => {
+    const { message, id, payload } = event.data;
+
+    if (event.origin === window.location.href || id !== this.id) return;
+
+    switch (message) {
+      case 'START_MONETIZATION':
+        return void this.sendStartMonetization(payload, true);
+      case 'RESUME_MONETIZATION':
+        return void this.sendResumeMonetization(payload, true);
+      default:
+        return;
+    }
+  };
+
+  /** @throws never throws */
+  private async checkLink(link: HTMLLinkElement) {
+    if (!(link instanceof HTMLLinkElement && link.rel === 'monetization')) {
+      return null;
+    }
+    if (link.hasAttribute('disabled')) {
+      return null;
+    }
+
+    const walletAddress = await this.validateLink(link);
+    if (!walletAddress) {
+      return null;
+    }
+
+    return {
+      link,
+      details: {
+        requestId: crypto.randomUUID(),
+        walletAddress: walletAddress,
+      },
+    };
+  }
+
+  /** @throws never throws */
+  private async validateLink(
+    link: HTMLLinkElement,
+  ): Promise<WalletAddress | null> {
+    const walletAddressUrl = link.href.trim();
+    try {
+      checkWalletAddressUrlFormat(walletAddressUrl);
+      const response = await this.message.send('CHECK_WALLET_ADDRESS_URL', {
```

Comment:
Let's rename `CHECK_WALLET_ADDRESS_URL` to something like `GET_WALLET_ADDRESS_INFO`.
```suggestion
      const response = await this.message.send('CHECK_WALLET_ADDRESS_URL', {
```